compiler/optimizer/VPHandlers.cpp

```diff
@@ -6931,7 +6931,7 @@ TR::Node *constrainIabs(OMR::ValuePropagation *vp, TR::Node *node)
          {
          int32_t value = low;
          if (value < 0)
-            value = -(uint32_t)value;
+            value = -value;
```

Comment:
This seems to have been introduced in #2711. @jdmpapin FYI. I think the change here is what we want. Could you please review this one? Here was the warning:

```
warning C4146: unary minus operator applied to unsigned type, result still unsigned
```

Reply:
The cast I introduced there was definitely intentional:
> Additionally, these handlers will now cast to unsigned before negating when computing the constraint for a constant, in order to ensure that there is no signed overflow.

Removing the cast will cause the (signed) negation to overflow if `value` is `INT32_MIN`, and that overflow is undefined behaviour (though it will probably happen to work fine anyway). And it's similar in `constrainLabs()` just below when `value` is `INT64_MIN`.

I think this is a bad warning in these two cases. It doesn't matter at all that the result is still unsigned, because the only use of the result is to immediately reinterpret it as a signed integer of the same size. And the change that the warning seems to encourage (i.e. removing the cast) reintroduces UB that the cast was introduced expressly to avoid.

It's possible though that appeasing the compiler will still result in cleaner code than suppressing the warning here with an explanation why. In that case, we can remove the cast, but we need to exclude the `INT32_MIN`/`INT64_MIN` cases, e.g.
```
if (value < 0 && value != INT32_MIN)
   value = -value;
```
Because negation doesn't do anything to the minimum value, it's fine to skip it.

Reply:
Thanks for the explanation. I've implemented the fix in a new commit so it is isolated. Fixed in 50d0dbd.